### PR TITLE
[Experimental] Direct stepper chunk support #2

### PR DESCRIPTION
**Pull request by colinrgodsey**
Continuation of #7012. Big thanks to @thinkyhead for helping me clean this up. Still more work to do, but it's at least updated and did a quick test to make sure it runs 👍 

[...]

This feature addition allows an external device to concurrently upload chunks of 1024 steps to the device, and trigger their sequencing by using a new G-code command (currently, 'C0'). The protocol for updating chunk buffers themselves is a binary protocol that starts a packet using the control character '!'. This character is not used elsewhere in g-code (input anyways), and allows low-level processing of the serial sequence- enabling buffering independent of the command parser (all handled in the ISR).

C0 command format:
_C0 **I**[chunk start index] **R**[number of chunks, defaults to 1] **S**[steps per second]_

The execution of the chunks is done by extending the Marlin block format with a field and flag that lets it execute the buffered chunk instead of looking for the normal trapezoid related parameters. The step speed is configurable, I've had success with 10k-30k steps/s, ~~although 30k seems to starve the temperature ISR causing runaway errors~~.

[...]

My external test planner: https://github.com/colinrgodsey/step-daemon

RepRap docs: http://reprap.org/wiki/Stepper_Chunks

More details in #7012.

### CODE REVIEW

File: Marlin/stepper.cpp [hunk 1]
```diff
@@ -832,6 +868,137 @@ void Stepper::isr() {
...
+      #endif
+    #endif // !ADVANCE && !LIN_ADVANCE
+
+    //if last step of this chunk (manual modulo)
+    if((step_events_completed16 & 0x3FF) == 0x3FF) {
+      //release chunk so it can be filled again
+      chunk_response[current_block->chunk_idx] = CHUNK_RESPONSE_NONE;
+
+      //keep iterating until we eventually reach our total step_events_completed
+      current_block->chunk_idx = (current_block->chunk_idx + 1) % (NUM_CHUNK_BUFFERS - 1);
```
Comment (colinrgodsey) [inline Marlin/stepper.cpp]: hmm i think i confused a modulo and a bitwise &... not sure why im subtracting one here.

File: Marlin/chunk_support.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+  //concurrently keep up with response counter, should auto roll over
+  while (chunk_respond_busy_done != chunk_respond_busy) {
+    SERIAL_PROTOCOLLNPGM("!busy");
+    chunk_respond_busy_done++;
+  }
+
+  //loop through every bucket, start 1 after the index of the last response we gave
+  for(uint8_t i = 1 ; i < (NUM_CHUNK_BUFFERS + 1) ; i++) {
+    const uint8_t idx = (last_response_index_start + i) % (NUM_CHUNK_BUFFERS - 1);
```
Comment (colinrgodsey) [inline Marlin/chunk_support.cpp]: hmm same thing here. I should maybe just change these to &'s and force the pow2 requirement.

Comment (thinkyhead) [reply Marlin/chunk_support.cpp]: Yes, simply enforcing powers of 2 will work. Just add the necessary test to `SanityCheck.h`.

**Comment by colinrgodsey** [comment]
Pushed some optimizations and cleanup that have been going good so far.

No sign of the temperature ISR issues anymore. I got rid of the per-tick updates to count_position and only do them every segment (8 ticks), so I cut the amount of 64-bit long emulation in 8 there. Seems to work really well, getting a lot more consistent sound out of the thing, and 25k step/s works great. 

Finishing up my first full test print with OctoPrint + StepD finally (and these new Marlin changes). Frog model at 120mm/s with tons of curves, curved surface infill too. Constant 25k step/s (haven't tried 30k again). 500kbps USB (not really needed, but works fine). Atmega2560 @16MHz (MKS Base v1.4). 

Zero stutter at all from OctoPrint on this model, which it normally would. Posting the video mostly just for some "proof" that it actually functions, but also trying to showcase the full range of noise. There's one wave table (or combo?) that seems to produce an interesting sound, but for the most part my retraction and z-hop causes the most noise, which it always has.

https://youtu.be/vQWB500AFjg

frog1

frog2

File: Marlin/Configuration_adv.h [hunk 3]
```diff
@@ -755,6 +757,28 @@
...
+ * direct stepper buffers (chunks) instead of G0/1 movement commands.
+ * This frees up processing power on the Marlin device and allows
+ * a higher consistent step rate.
+ *
+ * This is useful when used with an "external planner" that
+ * acts as the USB host and controls the GCode stream while
+ * producing and triggering direct stepper chunks using the
+ * described protocol:
+ *
+ * <!insert links to docs here!>
```
Comment (colinrgodsey) [inline Marlin/Configuration_adv.h:773]: @thinkyhead do you know where would be a good place to write a draft specification for this support? ill have to dig around the reprap wiki and see if theres a good dumping ground somewhere, but that might not be the right place...

EDIT: actually, that looks like it might be the right place

**Comment by colinrgodsey** [comment]
also just finished my first print with octoprint+marlin+stepd all on the RPi 3. Works great, zero stutter. Both services taking less than 10% total CPU combined on the device.

**Comment by Roxy-3D** [comment]
And...  does this mean all current GCodes and MCodes need to give up any 'C' parameters they have?
After all...   I was told again and again that GCode can not have ANY REPETITION of a letter on a given line.   
```cpp
// Handle a known C, G, M, or T
  |   | switch (parser.command_letter) {
  |   | +
  |   | +    #if ENABLED(CHUNK_SUPPORT)
  |   | +
  |   | +      case 'C': switch (parser.codenum) {
  |   | +        case 0:
  |   | +          gcode_C0();
  |   | +          break;
  |   | +      }
  |   | +      break;
  |   | +    #endif // CHUNK_SUPPORT
  |   | +
  |   | case 'G': switch (parser.codenum) {
```

**Comment by colinrgodsey** [comment]
@Roxy-3D ahhh, ok i didnt really think about that. probably makes great sense for simplifying parsers and such. it can be another G-code

**Comment by Roxy-3D** [comment]
I actually don't really care...   G29 and G26 are not going to give up the use of their 'C' parameter.   But it seems a fair amount of the arguments against me using G and M as easy to remember parameters just got thrown in the trash heap.

**Comment by colinrgodsey** [comment]
hah, yea i feel like thats a whole debate i should just stay out of ;) either way, it's easily changed. ill see if any other input trickles in. I honestly expected that to be the first point of contention.

EDIT: @Roxy-3D hah, just noticed the fire storm going on in #7131. yea, g-code is pretty janky

File: Marlin/Configuration_adv.h (hunk 3 above)
Comment (thinkyhead) [reply Marlin/Configuration_adv.h:773]: I agree. Use the RepRap wiki. All kinds of draft specifications there!

**Comment by thinkyhead** [comment]
@colinrgodsey Given that this is a "proprietary extension" that you might say "bypasses" G-code, and is not widely supported (yet) by slicers and hosts, I'm not worried about the extension of G-code for this specific feature while it's being evaluated. However, I agree with Roxy that we should pick wisely. So, I propose for this extension the initial code "letter" should not be a letter at all, but we should use another character, such as one of these: `!@#$%`

As we are past the 1.1 release and have mailed that package, I feel that Marlin 2.0.x can be more "outside the box" with its protocols. I think that insofar as we have a G-Code protocol, it should try to adhere to the CNC methodologies. But, for binary, compressed, or otherwise optimized protocols like Stepper Chunks, we should feel more free to experiment and settle on what works best _technically_.

Obviously, we can provide pre-processor scripts to convert normal G-code into chunks or binary, or any mixture…

**Comment by colinrgodsey** [comment]
Small update from yesterday. Did a test print at 30k ticks/steps per second all from the RPi, worked great, and no sign of temperature ISR issues. Today I'll make an attempt at 40k, but mostly just for fun.

Also wanted to showcase some of how step-daemon was working on the RPi 3 alongside OctoPrint:

stepd top

So thats 30k ticks/s, with full per-segment bed leveling. Really not using a ton of CPU, still plenty of space for timelapses or whatnot, and the multi-threaded pipeline for step-daemon happily chugs away on all cores.

**Comment by colinrgodsey** [comment]
@thinkyhead I'm already using '!' as a submission control character, suppose I could use any of those other ones for 'triggering' the buffers. Hmm and yea, I agree you would never see this protocol in any g-code file, nor would you be able to really trigger it manually. 

Not a huge fan of the "proprietary" description tho, although it is definitely a non-standard transmission protocol. I'd love for this to be used... wherever, hoping to get a full proposal up on RepRap in the next day or two. Just want to verify that I don't benefit from this lol This doesn't seem like an area ~~of machining~~ where people go to get rich ;)

EDIT: sorry, reflecting too heavily on @bobc 's comments on the first PR

**Comment by colinrgodsey** [comment]
@thinkyhead ah, just saw your edits. Yea, this extension could technically be done entirely from the file. I was fiddling around with a web version of step-d that could do all the conversion just from the browser (normal JS app). Would be easy to have it write back to a file, but there's some other semantics I need to think over.

Marlin 2.x makes sense, definitely interested in taking a more technical deep dive with whoever wants to get involved. I've been really pleased with the results from this proof of concept so far, but there's no real rush or demand for this yet. If step daemon (or some other external planner or preprocessor) really showcases its worth then there might be, but that's not realistically going to happen soon without some sort of solid firmware support and candidate spec that's beneficial on a more fundamental level. 

But at this point, I've at least gotten over my OctoPrint stutter and can print reliably at 120mm/s (at 30k/s). So I'm happy for now.

**Comment by colinrgodsey** [comment]
started a reprap article for the actual protocol, could be a good place for high-level protocol talk: http://reprap.org/wiki/Stepper_Chunks

**Comment by davidelang** [comment]
arriving at the discussion a bit late, but I've read through the docs linked, and skimmed through the prior issue. I've also been looking at how grbl and maslow handle motion.

I think that wave tables are a bad option for the communication, they are going to be large, and one of the limiting factors is the communication speed (serial port)

Thinking about the requirements to define the motion, grbl has good information about the output of their planner, look at around line 172 at https://github.com/gnea/grbl/blob/master/grbl/stepper.c

They break things down to ramp-up, constant speed, ramp-down sections, as the output of their planner. That seems like a MUCH more compact representation than wave tables, while still doing all the hard math on a faster external system.

**Comment by davidelang** [comment]
Also, if you are going to do an external planner, you should also figure that you can have more than one board driving things.

This means that you need a way to sync multiple boards together (since you cannot simultaneously transmit a 'go' signal vial serial or network).

I would suggest something like what I2C and USB use, have a common line that has a pull-down (or up) resistor on it, and each device on the bus pulls it the other direction shortly after they start a sequence. This allows you to have a 'sync/start' command that releases the local pull up on the line and waits for the line to go low. Once all devices have hit the sync/start command, and the master has also released the line, it will drop and all boards can start simultaneously.

As long as the motion commands send to the various boards are supposed to take the same amount of time, and you put a sync in every few seconds, the drift in the time between the different boards canb e kept to an acceptable level.

If we are going to define a new protocol for controlling CNC machines, let's make it be something that can control any type of machine with any number of axis rather than something limited to 3D printing with 1-2 extruders. Let's design something that would support having the 4-filament extruder one one controller board with a 4+ axis motion control system on another controller board as an example, or a 5-axis milling machine with another axis or two to control tool selection.

**Comment by colinrgodsey** [comment]
@davidelang ah, so the wave tables for the CH-4x4-128 format are actually fixed tables, they never get sent over the wire. They're actually totally optional, but for that particular format the wave table is rather small (128 bytes) and used just to avoid having to do the Bresenham line math in the stepper ISR. For other formats, the wave tables might not make sense, or might not have any use (like 1-tick segment encodings, there's no lines to interpolate over). I might also go back to looking at Bresenham lines for this format, but I'm not sure I can do it with pure 8-bit math (the stepper ISR has to be really lean to keep up with the constant high tick rate, 30k/s is the highest ive gone so far). The chunk format should have a constant data transmission rate for a certain tick rate (unless we start talking about compression, which is something to discuss down the road). So for 30k tick/s, there's a constant 30 chunks/s, about 61kbps over serial.

That encoding basically breaks all moves down into 8-tick 'lines' or segments. So each 4 bits of the encoding describes a line that's either... +1 steps over 8 ticks, -2 steps over 8 ticks, etc up to +/-7 steps over 8 ticks. Or basically it describes the velocity for that stepper for those 8 ticks. The wave tables are there in that format just for a quick lookup so the device knows how to distribute the... +3 steps over 8 ticks, etc. 

Transmitting pure "GRBL style" trapezoids (physics-adjusted movements) is definitely something i considered, in fact I always wondered why there wasnt just an extension to G0 commands that let you do that already. In this case, we could save total machine cycles by limiting the amount of planning math, but the stepper ISR is still just as busy as it was (preventing higher consistent tick rates). Plus with the chunk format, you're able to do more non-trapezoidal work externally, like advanced bed leveling or linear advance, where you need influence at a per-step level (technically per 8 ticks with CH-4x4-128), and not just per-trapezoid. For example, step-daemon does bed leveling per segment (8 ticks) vs per trapezoid like its generally done on the device. One reason I started on all this was because I wanted to do a more 'realistic' simulation of linear advance that would be basically impossible on an 8-bit machine. 

The multi-device thing is very interesting, haven't really put much thought into it, but you're definitely right. I think there should be some way for controlling multiple devices and syncing them, which should be very possible with the external planner.

As far as multiple steppers, the formats should have a pretty standard way of modifying them, plus should support a few "standards" that are optimized for various machines. So for CH-4x4-128 (which is 4-bit segment, 4 motors, 128  segments) could be expanded out to be CH-4x8-64 (4-bit segment, 8 motors, 64 segments) or some other variant rather easily. The one I provided is an "optimized 4 motor format for 8-bit control boards". The format, chunk size, wave tables, etc. were all done with 8-bit optimization, 4 motor, and high tick-rate in mind. But there should definitely be some standards for other cases. So for example with Marlin, you would want to compile the firmware with the format that works best for your configuration. 

One example I can throw out there is webcams. Webcams support certain color spaces and encodings for their video output. Most webcam host software is generic enough to support most or all of these formats and color spaces. The device itself however is normally very limited, and generally only supports the formats etc that work best with the hardware, while maintaining a set quality and data rate. 

So what I basically see happening, is that your host device can probably support any format out of the box, at any time. But the control device itself will probably be limited to just a few formats that the hardware can handle, and depending on desired tick-rate, motors, etc.

EDIT: I also understand this a *hard* departure from GRBL style trapezoids. GRBL is based around moving in fixed-speed lines, but adjusting the time domain (tick rate) to simulate acceleration. This implementation flips that and keeps a constant tick rate but modifies the segments (essentially velocity) to simulate the same physics. At this point I can confirm the physics are just as accurate either way, while maybe reducing total noise with the high-rate chunk format.

**Comment by davidelang** [comment]
Ok, I don't understand what the wave tables are. I thought they were the specific step pattern to be used to drive the steppers for all axis at once. That would require changing the wave table every time the relationship between the different axis changes.

so if your method of syncing the different axis is not to change the wave tables to reflect the different speeds of each axis, how do you sync them?

**Comment by colinrgodsey** [comment]
I need to stop writing a small book every time i respond, too much text lol

@davidelang  Definitely appreciate the feedback so far, this will be crucial in making the RepRap article not horribly confusing. I'll try to boil it down a bit more directly:

For example, we have: a chunk tick rate of 30k tick/s (_ticksPerSecond_). That means the Marlin stepper ISR is processing steps at a fixed and constant rate of 30k times per second. Keeping consistent with the above format, we have 4-bit 'segments', but we'll just assume this chunk format only encodes one motor/axis, '_X_'. The chunk has a series of 4-bit segments in it, each segment encodes a number of steps (I'll call this 'delta X', or _dX_) to take over the next 8 ticks. A 4-bit value can express 16 different numbers, and here it's encoded as "dX + 7". So a dX of -2 would be encoded as 5, -3 would be 4, 1 would be 8. This gives us a range of +/- 7 (0 to 15,  minus 7, gives -7 to 8) for each segment. So we can describe anywhere from -7/8 steps/tick, to 7/8 steps/tick. The 8/8 encoding is not used because there is no -8/8 (that would require 17 values). **TANGENT**: _This unused value could also be very important in other iterations of this format, and could provide a control byte used for further compression or something down the road. It offers an unused "magic number"._

So to take that delta value, dX in this case, and get that actual step rate for that segment, you would do:

_dX_ / 8 * _ticksPerSecond_ = steps/s for _X_

So in GRBL and Marlin during movement, that per-axis step rate is acted on using something called 'Bresenham lines', which is a way to break the real-number step rate into integer steps. Same algo is used to draw a line on a computer monitor, as far as deciding where to place pixels on the line. Same concept applies here, but our discrete pixels are _ticks_ in this case. 

Now, while Marlin and GRBL have to normally deal with larger lines, and a wider array of slopes/rates, this encoding does not. In fact, there's only the 15 rates possible, -7/8 to 7/8. This is possible because of the very small segment size. And with slower moves, you'll generally be flipping back and forth between 2 of the low rates (any rate below 3750 steps/s will flip between _dX_ of 0 and 1). Same with real-numbered rates, if my axis was moving at 2.5/8 * _ticksPerSecond_ steps/s, the segments would appear as alternating rates of 2/8 and 3/8, or _dX_ of 2 and 3.

So, this is where the wave tables come in. This device needs to routinely figure out how to distribute _dX_ steps over 8 ticks. And because there's only 15 possible values for dX, we can skip a ton of math and just make a static table that says how to distribute _dX_ steps over those 8 ticks. So if we want to think of dX as an audio frequency (which it sorta is) we can just look up the wave form in the tables. 

Now, as far as accuracy, the 4-bit (per 8-step) segments do suffer some precision loss, but only at high-speed direction changes. Because high-speed direction change is avoided anyways (jerk, acceleration), it's not a big issue. Also the 8-step resolution is generally hidden within the micro-stepping curve (assuming 8 or more micro steps), which generally produces a non-linear transition anyways. The end effect is a compromise that's pretty consistent with 3d printing: more accurate directional change at lower speeds.

TL;DR- The 4-bit segments are basically a low-bit encoding of the velocity, that can change every 8 ticks.

**Comment by Roxy-3D** [comment]
> I need to stop writing a small book every time i respond, too much text lol

No!   This is a complicated new area.   The only way to get people bought in and supportive is to have all the details there.   And probably answer the same questions that get asked again and again...  It is always hard when you are trying to break new ground.

**Comment by colinrgodsey** [comment]
@Roxy-3D roger that! I'm ok with re-explaining things, just not the best 'presenter'. But you're absolutely right, and this will definitely help lead to a more succinct writeup on the wiki.

**Comment by davidelang** [comment]
In many ways, this is a far better place to discuss these things than the reprap wiki (difficulties in getting accounts there for example), long explinations are needed.

so the wave tables just convert the desired rates to the individual steps.

So what is the protocol for sending the desired rates? at 30K tics/sec / 8 tics/rate, this means that you need to send 4k rates/sec to the device (per axis). This is unsustainable, so how do you compress these streams?

**Comment by colinrgodsey** [comment]
@davidelang So the current format described above (CH-4x4-128), does 4 axis, 4-bit segments (1/2 byte) over 8 steps, 128 segments total (for each of the 4 axis), per chunk. Total of 256 bytes (0.5 * 4 * 128). Reduced, each chunk will encode 1024 ticks (128 * 8) worth of data for 4 axis, in that 256 bytes. Plus an additional 2 bytes needed for the transmission format, total of 258 bytes per chunk over the wire. Assuming we're at 30k tick/s, that's 7.56 KB/s, and finally 60kbps. 

Back when I was doing initial throughput testing, I was able to achieve about 60% of the effective line bitrate in "real" chunk transfer using the current protocol, which required processing responses, etc. So the above format, even at 40k ticks/s, would work just fine at 250kbps with decent headroom. I'm currently using 500kbps baud, but just because I can now that the stepper ISR is light enough to not be effected by the UART interrupts. This was done on an Atmega2560 based board in both of my tests (a simple arduino, and my real printer). I'll do a test at 250kbps just to confirm, but it honestly shouldn't matter. Neither bitrates are 'standard' as far as host devices go, and 500kbps runs with 0 issue on the device, so I'm finding it hard to justify recommending anything but 500kbps.

Since doing that above throughput test, I've also done plenty of optimization on the host software that should let it achieve higher actual throughput (the main bottleneck was latency related, while the host was waiting on a response from the device), although I have no real numbers at this moment. So far I've just been sticking with the original numbers and using them as a "worst case".

**Comment by davidelang** [comment]
do you override the default 64k buffer from the arduino serial library?

**Comment by colinrgodsey** [comment]
@davidelang yep! The chunk buffering sits straight in the ISR and directs the chunk data to the right buffer. It's sort of like a lazy DMA, so you can think of the host as actually writing binary data directly to dedicated buffers. The host starts sending chunk data, designated by the right control character (currently '!'), and this starts redirecting the data to the buffer. Once the chunk is buffered (there's also a checksum), it queues up a response that has the "chunk index", or chunk number, back to the host. That number is then used later when its time to "execute" the stepper chunks. 

A good analogy here might be how you would buffer data to a graphics card with... OpenGL or something. Your software says it needs a buffer, and sends the data down the pipeline, and eventually is given a handle back that it can use later with other instructions to the GPU (actually the handle comes first in the OpenGL case.... but we'll ignore that). Either way, it's in the hosts advantage to get the 'big data' onto the device well before its needed, and as efficiently as possible. And the small handle allows it to execute/consume the data on the device quickly as possible once its buffered.  

EDIT: the UART ISR works as normal outside of the chunk protocol. So if it's not "buffering chunks", the data gets sent down the normal hardware serial pipeline and functions like it always did. 

EDIT2: Also, Marlin currently uses a custom version of the HardwareSerial library imported in the code base. So modifying it was easy, and didn't involve any departures from the normal build process.

**Comment by davidelang** [comment]
so with this approach, you need 514 bytes of serial buffering (the bock you are working on, and the next block so you don't have to pause after hitting the end of the first one)

This also means that the system sending you data had better not pause for more than ~1/60 of a second (you need to send a new chunk of data every 1/30 of a second and it takes ~1/60 of a second to send the data), that's getting to be a pretty tight window.

It seems to me that a slightly higher level protocol would be better. If you are printing small things where you have very few straight segments, you won't see the inefficiencies, but with larger scale systems, you will have more straight lines that can be compressed significantly. Yes, it will eat up a little cpu on the arduino to uncompress things, but I'd bet that this will end up comparing reasonably with the need to use that cpu to transfer additional data.

Also, there are getting to be more smart devices out there that take velocity/distance instructions, not just step/direction instructions. mecharduino, pthat, odrive are just a few of the options out there. If we're going to create a new protocol to offload the planning from the motion controller board, we should look at a protocol that has the ability to be useful across different systems.

Another issue that I've seen fly by in my e-mail talked about some motor control that used an i2c interface to the motors.

With the easy availability of high resolution position encoders, coupling a high-res encoder (up to a few thousand positions/rev) with a plain DC motor, servo, or brushless DC motor is a viable option (especially when high power or high speeds are involved)

Open-loop control of stepper motors has enabled the current generation of CNC machines, but steppers seem to have difficulties scaling to higher power and higher speeds. Meanwhile brushless DC motors have been benefiting from the quadcopter craze and are available in power and speed options that dwarf steppers, the problem has been in controlling them. We are now seeing the barriers in this control start to fall, and while I don't expect to see them in a <$300 3D printer, I do expect to see them in larger CNC devices.

(sorry for the long-winded reply :-)

**Comment by colinrgodsey** [comment]
> (sorry for the long-winded reply :-)

You're in the luck! Long winded replies are in season ;)

So the current implementation for Marlin I've been testing with can actually buffer 16 chunks. A massive 4k! At 30k/s chunk tick, its a full half second of moves. I was able to reduce the planner 'block' buffers on Marlin to only 8. The block buffers take up a ton of memory, and mostly contain planning/accel/etc related parameters, so it was a great place to steal RAM from. We still need the block buffers as Marlin is still capable of planning, but for high-speed chunk related movements we only need a few specialized "blocks" that are capable of playing the chunk buffers (a "chunk block" can execute many chunks, currently my software limits it to 5 chunks per block). So with only 16 chunk buffers, at most (currently with my host software), we'll need 3 "blocks" to execute 15 chunks. The last part is getting a bit into the weeds, but basically the chunk execution still lies within the normal Marlin planner operation, but we're able to significantly reduce the planner buffer and overhead needed. 

So, with the max 1/2 second buffer, latency is still definitely an issue. The host device can really only afford so much missed transfer time before its behind. I did all of my initial tests from my Mac, and only heard "stuttering" at 1 or 2 discrete moments (it just dwelled a sub millisecond and moved on) over a 3 hour print. One of those moments was when I accidentally closed chrome and had to re-open it. The other time was related to some other system event I can't remember. But given that my chrome eats 70% of my CPU while restoring tabs, I would still call that a win. My test planner is also written in Java which is notorious for pauses... but thankfully on JVM 1.8 I've yet to monitor any pause thats more than 2ms. This definitely requires some smarter host logic, but it is entirely possible. 

My RPi3 also didn't experience any stutters when testing the host (this was also using OctoPrint as the g-code source, via my host software, so was octoprint->stepd(host)->USB->marlin), until I forgot I was using it and ran a system update on it. I only heard one blip from that though still. After that, I set priorities on the host software (was already doing this for octoprint) and that fixed the issue entirely. I can't get it to miss a beat no matter what I try now. Given the fact the host is a semi-realtime system, its entirely reasonable to give it priority over everything else in the system. It runs at less than 10% total CPU usage, so it won't starve time from anything in any meaningful way. 

I agree about the compression, and I think that's a logical next step. But for this phase, I wanted to avoid compression entirely just for performance judgement and "proveability". The current format has the same data transfer needed per second, no matter what (unless its not moving). It occupies both the best and worst case scenario, whereas compression gives you a range between those (might stutter on small moves, but is fine on large moves) and different performance profiles to consider. I'm already thinking about compressed formats, and have some good ideas there. Tons of options. 

I've yet to really look into the other types of smart devices out there and their corresponding formats. It would be great if there was a suitable common format for sure... I'll have to see exactly what they're doing there. My main goal here was to provide an "advanced processing" open-source solution that uses rather low-end devices that users probably have access to: a Marlin compatible control board, and an RPi. I'll look into these other devices though, there's plenty of smart people in this landscape and I'm sure some great solution, and as long as none of them rely on anything too "complicated" it could possibly work. 

Hmmm, as far as the brushless stuff... that kinda blows my mind right now lol. I knew the accuracy and control has gone through the roof on those things, but how you handle accuracy and repeatability... is currently beyond me. 

It's very interesting to hear about the move away from steppers, I've assumed that was bound to happen eventually. I've always wondered about options like laser telemetry or something for analog motor positioning.... and I guess in some ways this could lead to support of something like an "external planner" like this. My current test planner (step-daemon) handles everything as cartesian 64-bit double floats and time deltas until the very end where it needs to produce step sequences. Substituting some real-number format using floats would be super easy provided the target device can handle it, vs other planners that are step oriented from the beginning of planning. 

Well... I believe I have some research to do tonight.

**Comment by davidelang** [comment]
On Wed, 12 Jul 2017, Colin Godsey wrote:

> Hmmm, as far as the brushless stuff... that kinda blows my mind right now lol. 
> I knew the accuracy and control has gone through the roof on those things, but 
> how you handle accuracy and repeatability... is currently beyond me.

these sensors can give you 2400 steps/rev

http://www.ebay.com/itm/360-600P-R-Photoelectric-Incremental-Rotary-Encoder-5-24V-AB-Two-Phases-Shaft/142133029511?ssPageName=STRK%3AMEBIDX%3AIT&var=441209388746&_trksid=p2060353.m1438.l2649

these can give you 4000 steps/rev

http://www.ebay.com/itm/OMRON-E6B2-CWZ6C-Rotary-Encoder-1000P-R-New/252856681203?ssPageName=STRK%3AMEBIDX%3AIT&_trksid=p2060353.m2749.l2649

There are some others out there that can give you 16k steps/rev

(take the pulses/rev and multiply by 4 to account for the two different pulse 
trains that are out of step with each other, giving you 4 state changes per 
pulse)

If you know where you are from reading the encoder, you can adjust the power to 
the different coils to move you go a given position, no matter what the extenal 
forces are on the motor (which means you can maintain the position, even in the 
face of changing forces)

This is like microstepping a stepper, but with feedback so that you know where 
you really are rather than just predicting the position based on non load.

> It's very interesting to hear about the move away from steppers, I've assumed 
> that was bound to happen eventually.

to be clear, this is my prediction based on the technology I'm seeing show up in 
opensource projects.

a few years ago there was the makesmith CNC for <$200 that used servos for 
motion with encoders to detect the position of the shaft.

Currently there is the Maslow CNC (same folks) that uses 12v DC motor/gearbox 
setups for positioning, with an encoder to track how far the motor has turned 
(8148 steps/rev of the final output shaft)

the odrive project is a controller that uses the first encoder I listed above 
combined with a high power brushless DC hobby motor to provide almost 2.5Kw of 
power (per axis, but can't sustain this on both axis simultaniously) with a 
resolution of 2400 steps/rev

the pthat drives traditional steppers, but it does so at very high speed (over 
100K steps/sec)

the mecharduino is a controller board that hooks to a standard stepper (in the 
current iteration), but that has an encoder to detect the actual position of the 
stepper, so it can get full torque at microstep accuracy and compensate for 
loads that are large enough to move the stepper.

These are just the ones I've run across and remember off the top of my head.

**Comment by colinrgodsey** [comment]
Wow, this pthat thing is pretty cool. I always felt like there was missed opportunity for using the RPi for real time tasks. Linux isn't a real-time OS out of the box, but these days it takes almost no work to boost a process to real-time priority. You barely even need drivers anymore, you can do so much from userspace processes (although you get better CPU use with kernel drivers due to some semantics). Hell, I even run OctoPrint as real-time on my RPi. If it bugs out it locks the thing up... but big deal. Power cycle it. 

I forgot about position encoders! Yea, that's definitely a larger number of steps to encode. Something like that would need some sort of compressed chunk format to function. There's a lot of entropy that can be safely reduced out there, and probably little chance that... you'll be changing directions more than once every X hundred steps. It's definitely something i want to explore soon.

I'll have to do some more digging around and see what the protocols are like for these things.

**Comment by davidelang** [comment]
pthat probably has the best protocol definition in place.

**Comment by colinrgodsey** [comment]
update: this feature is not dead, my time availability the last few months has just dropped to the floor, hard. aiming to pick this back up next month

**Comment by thinkyhead** [comment]
It'll probably be not-too-hard to bring up to date. This feature doesn't rely too much on other stuff.

**Comment by manotroll** [comment]
Is the project still open?

**Comment by w1ebr** [comment]
Hello! Are you still working on this? Thanks!

**Comment by colinrgodsey** [comment]
I'm looking to get this branch updated again very soon. Had a busy.. half year or so. I still have more work to do on the wiki, but planning to go full-force again with this soon.

**Comment by thinkyhead** [comment]
At this point Marlin 1.1.x is end-of-life, so this PR will need to be re-done starting from `bugfix-2.0.x` and targeting that branch. Sorry for the inconvenience!

**Comment by GMagician** [comment]
I'm wondering why this has been lost..it was a good idea helpful for laser systems. I don't need it, but I think Marlin lost a good feature.

This will remind me a famous sentence in Blade-runner:
> [cut]. All those moments will be lost in time, like tears in rain. Time to die.

**Comment by thinkyhead** [comment]
There are not too many changes here. Bringing them over to 2.0.x should only take a day. The final PR would have to be accompanied by a pre-processor to convert G-code to C-code for various kinematics and planner settings.

As for laser systems, I'm working on code to change PWM DC faster and in sync with the stepper ISR which should help a lot. After that, the laser raster buffer feature of certain other firmwares also needs to be ported over so that we can do raster scaling and mirroring on the fly and other cool laser tricks.

**Comment by colinrgodsey** [comment]
revived this in https://github.com/MarlinFirmware/Marlin/pull/17853
